layers/+lang/mercury-layer/packages.el

```diff
@@ -0,0 +1,77 @@
+;;; packages.el --- mercury layer packages file for Spacemacs.
+;;
+;; Copyright (c) 2012-2018 Sylvain Benner & Contributors
+;;
+;; Author: Ludvig Böklin <ludvig.boklin@protonmail.com>
+;; URL: https://github.com/syl20bnr/spacemacs
+;;
+;; This file is not part of GNU Emacs.
+;;
+;;; License: GPLv3
+
+;;; Commentary:
+
+;; See the Spacemacs documentation and FAQs for instructions on how to implement
+;; a new layer:
+;;
+;;   SPC h SPC layers RET
+;;
+;;
+;; Briefly, each package to be installed or configured by this layer should be
+;; added to `mercury-packages'. Then, for each package PACKAGE:
+;;
+;; - If PACKAGE is not referenced by any other Spacemacs layer, define a
+;;   function `mercury/init-PACKAGE' to load and initialize the package.
+
+;; - Otherwise, PACKAGE is already referenced by another Spacemacs layer, so
+;;   define the functions `mercury/pre-init-PACKAGE' and/or
+;;   `mercury/post-init-PACKAGE' to customize the package as it is loaded.
+
+;;; Code:
+
+(defconst mercury-packages
+  '((metal-mercury-mode :location (recipe
+                                   :fetcher github
+                                   :repo "ahungry/metal-mercury-mode"
+                                   :commit "99e2d8fb7177cae3bfa2dec2910fc28216d5f5a8"))
+    (mercury-mode :excluded t)
+    flycheck
+    (flycheck-mercury :requires flycheck)
+    smartparens))
+
+(defun mercury/post-init-flycheck ()
+  (add-hook 'metal-mercury-mode-hook 'flycheck-mode))
+
+(defun mercury/init-flycheck-mercury ()
+  "Initialize flycheck-mercury"
+  (use-package flycheck-mercury
+    :defer t
+    :init (require 'flycheck-mercury)))
```

Comment:
Never use require in an init block, this renders your defer use-less.
Rather use a hook to load flycheck with mercury mode.

Reply:
Thanks for the feedback. I'm not very proficient in elisp, would you mind specifying in more detail where I should add the hook? In `mercury/init-metal-mercury-mode`?

Reply:
Sure, you should add it to the init code of flycheck-mercury something among the line of
```elisp
(defun mercury/init-flycheck-mercury ()
  "Initialize flycheck-mercury"
  (use-package flycheck-mercury
    :defer t
    :init (add-hook 'metal-mercury-mode-hook Your-init-function)))
```
The idea is not to require the package but rather call one of the functions to autoload this package. Which function this is depends on the package. 

After checking the package in question I have seen that there is no autoload defined in there, so for now I would use the following code to init the package:
```elisp
(defun mercury/init-flycheck-mercury ()
  "Initialize flycheck-mercury"
  (use-package flycheck-mercury))
```

This should load the package immediately, we can later tweak it by adding autoloads from the layer.